lib/opte/src/ddi/sync.rs

```diff
@@ -409,3 +427,327 @@ impl<T> KRwLock<T> {
+                        dur.as_nanos() as clock_t,
+                        time_res_t::TR_NANOSEC,
+                    )
+                } {
+                    -1 => TimedWaitSigCause::Timeout,
+                    0 => TimedWaitSigCause::Signal,
+                    a if a > 0 => TimedWaitSigCause::Notify,
+                    _ => panic!("illegal return value for cv_reltimedwait_sig"),
+                };
+                (lock, cause)
+            } else {
+                let (lock, cause) = self.wait_timeout(lock, dur);
+
+                let cause = match cause {
+                    TimedWaitCause::Notify => TimedWaitSigCause::Notify,
+                    TimedWaitCause::Timeout => TimedWaitSigCause::Timeout,
+                };
+
+                (lock, cause)
+            }
+        }
+    }
+}
+
+#[cfg(all(not(feature = "std"), not(test)))]
+impl Drop for KCondvar {
+    fn drop(&mut self) {
+        unsafe { cv_destroy(self.cv.get()) };
+    }
+}
+
+pub enum WaitSigCause {
+    Notify,
+    Signal,
+}
+
+pub enum TimedWaitCause {
+    Notify,
+    Timeout,
+}
+
+pub enum TimedWaitSigCause {
+    Notify,
+    Signal,
+    Timeout,
+}
+
+/// A mutual exclusion mechanism which loans out access to a single
+/// internal token. This is used to ensure at most one thread is present
+/// in a critical section *without actively holding a [`KMutex`]*.
+///
+/// This is necessary for some kernel-level operations which will upcall or
+/// enter some other context in which it is unsafe to hold a [`KMutex`] or
+/// [`KRwLock`]. Any functions attached to the token `T` should denote whether
+/// these restrictions apply.
+pub struct TokenLock<T> {
+    // In future, this could be arbitrary (i.e., taking user-held
+    // context like a file descriptor). Similarly this could allow
+    // for re-entrancy, but this is tricky to square with `&mut` access
+    // to the inner `T`.
+    #[cfg(all(not(feature = "std"), not(test)))]
+    holder: KMutex<Option<NonNull<kthread_t>>>,
+    #[cfg(any(feature = "std", test))]
+    holder: KMutex<Option<ThreadId>>,
+    cv: KCondvar,
+    inner: UnsafeCell<T>,
+}
+
+impl<T> TokenLock<T> {
+    pub fn new(token: T) -> Self {
+        let holder = KMutex::new(None);
+        let cv = KCondvar::new();
+
+        Self { holder, cv, inner: UnsafeCell::new(token) }
+    }
+
+    pub fn lock(&self) -> Token<'_, T> {
+        let mut thread_lock = self.holder.lock();
+
+        while thread_lock.is_some() {
+            // Here, we expect that there is a very real possibility
+            // that the kthread in `thread_lock` is, itself, susceptible
+            // to being `STOPPED` (and thus deadlocking us) if we cannot
+            // also handle a SIGSTOP. We don't care whether we were awoken
+            // by the signal or the CV, only that our thread can also stop
+            // to allow a fork of the caller to complete.
+            thread_lock = self.cv.wait_sig(thread_lock).0;
+        }
```

Comment:
I think we probably want two different versions of `TokenLock::lock()` - one which is sensitive to signals (for use in ioctls, etc), and one which is not (for use in attach/detach).  The signal-sensitive locking function (lock_sig()?) could then return an error if `cv.wait_sig()` indicated reception of a signal, allowing the ioctl to emit `EAGAIN`, rather than potentially going into a furious loop around the `cv.wait_sig()` call if a signal is pending.